extensions/Worker.Extensions.SignalRService/src/Hubs/ServerlessHubOfT.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Azure.Core.Serialization;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Azure.SignalR.Management;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Azure.Functions.Worker.SignalRService
+{
+    public abstract class ServerlessHub<T> : ServerlessHub where T : class
+    {
+        protected ServerlessHub(IServiceProvider serviceProvider) : base(serviceProvider)
+        {
+        }
+
+        private ServiceHubContext<T> HubContext
+        {
+            get
+            {
+                var type = GetType();
+                var hubContextProvider = ServiceProvider.GetRequiredService<HubContextProvider>();
+                if (hubContextProvider.TryGetValue(type, out var hubContext))
+                {
+                    return (ServiceHubContext<T>)hubContext;
+                }
+                else
+                {
+                    throw new InvalidOperationException($"The serverlesshub {type.Name} is not registered using services.AddServerlessHub<T>().");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets an object that can be used to invoke methods on the clients connected to this hub.
+        /// </summary>
+        protected new IHubClients<T> Clients => HubContext.Clients;
+
+        /// <summary>
+        /// Get the group manager of this hub.
+        /// </summary>
+        protected new GroupManager Groups => HubContext.Groups;
```

Comment:
For all of these properties where the types line up between base class and this class, can `virtual` and `override` be used instead?

Reply:
Sure, `virtual` also allows the users to mock the impl and unit test it.